core/pta/attestation.c

```diff
@@ -0,0 +1,589 @@
+		dest_ptr += block_size;
+		src_ptr += block_size;
+	}
+
+	crypto_cipher_final(ctx);
+	*plain_size = enc_size - plain[enc_size - 1];
+
+	return res;
+}
+
+static TEE_Result get_sha256(uint8_t *dst, uint8_t *src, size_t src_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *ctx = NULL;
+	size_t digest_len = TEE_SHA256_HASH_SIZE;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		return res;
+
+	res = crypto_hash_update(ctx, src, src_len);
+	if (res)
+		return res;
+
+	res = crypto_hash_final(ctx, dst, digest_len);
+	if (res)
+		return res;
+
+	crypto_hash_free_ctx(ctx);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result acipher_sign(const uint8_t *msg, const size_t msg_len,
+			       uint8_t *sig, size_t *sig_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rsa_keypair key_pair;
+	uint8_t *d_ptr = NULL;
+	uint8_t *n_ptr = NULL;
+	uint8_t *e_ptr = NULL;
+	size_t d_size = 0;
+	size_t n_size = 0;
+	size_t e_size = 0;
+	size_t plain_priv_len = 0;
+	uint8_t *plain_priv = calloc(att_data.enc_priv_size, sizeof(uint8_t));
+
+	if (!plain_priv)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	// decrypt the AIK
+	res = decrypt_priv_key(plain_priv, &plain_priv_len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	// use it to sign the msg
+	res = crypto_acipher_alloc_rsa_keypair(&key_pair, 0);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = pub_asn1_decode(&n_ptr, &e_ptr, &n_size, &e_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = priv_asn1_decode(plain_priv, plain_priv_len, &d_ptr, &d_size);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(d_ptr, d_size, key_pair.d);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(e_ptr, e_size, key_pair.e);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(n_ptr, n_size, key_pair.n);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_acipher_rsassa_sign(TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256,
+					 &key_pair, // keys
+					 0,	// salt, default is 20
+					 msg,
+					 msg_len,
+					 sig,
+					 sig_len);
+	if (res)
+		return res;
+
+	free(plain_priv);
+	crypto_bignum_free(key_pair.e);
+	crypto_bignum_free(key_pair.d);
+	crypto_bignum_free(key_pair.n);
+	return TEE_SUCCESS;
+}
+
+/*
+ * encrypt a random number to test the RSA keypair received through the
+ * set_data call from the normal world.
+ */
+static TEE_Result check_attestation_data(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	// Decrypt the encrypted private AIK
+	size_t plain_priv_len = 0;
+	uint8_t *plain_priv = calloc(att_data.enc_priv_size, sizeof(uint8_t));
+
+	if (!plain_priv)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = decrypt_priv_key(plain_priv, &plain_priv_len);
+	if (res)
+		return res;
+
+	// get a random number
+	size_t rn_len = 32;
+	uint8_t rn[rn_len];
+
+	res = crypto_rng_read(rn, rn_len);
+	if (res)
+		return res;
+
+	// decode the pubkey (AIK)
+	struct rsa_public_key key;
```

Comment:
Please declare local variables at the beginning of a code block.

Reply:
Fix [here](https://github.com/OP-TEE/optee_os/pull/4011/commits/c40958f4ba9061e13ebccf202b316e0cb1fef030).